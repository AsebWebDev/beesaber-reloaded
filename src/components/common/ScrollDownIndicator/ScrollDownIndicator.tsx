import styled from 'styled-components';

import useIsScrollingDown from '@/sharedHooks/useIsScrollingDown';

const Indicator = styled.div`
    position: relative;
    width: 100%;
  
  ::before {
    animation: bounce 1s ease infinite;
    bottom: 4rem;
    color: #fff;
    content: '╲╱';
    font-size: 2rem;
    height: 1rem;
    left: 50%;
    letter-spacing: -1px;
    line-height: 4rem;
    margin-left: -3rem;
    opacity: 0.8;
    position: absolute;
    text-align: center;
    width: 6rem;
  }
  
  @keyframes bounce {
    50% {
      transform: translateY(-50%);
    }
`;

const ScrollDownIndicator = (): JSX.Element | null => {
  const { isScrollingDown } = useIsScrollingDown();

  if (isScrollingDown) return null;

  return <Indicator data-testid={'indicator'} />;
};

export default ScrollDownIndicator;
