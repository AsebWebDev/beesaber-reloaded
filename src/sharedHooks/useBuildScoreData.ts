/* eslint-disable no-console */
import { useEffect, useState } from 'react';
import { toast } from 'react-toastify';

import getAllScores from '@/api/services/helper/getAllScores';
import { useAppDispatch } from '@/store/hooks';
import { userIsFetchingData } from '@/store/reducer/appStatusReducer';

import type { ScoreData } from '@/../sharedTypes';

/**
 * Fetch and parse all score data of a user
 * @param {String} id - ScoreSaber id
 * @param {number} threshold - optional threshold for number of score pages (1 page = 8 scores)
 * @returns {ScoreData} score data object with recent scores, top scores and a list of
 * score hashes
 */
const useBuildScoreData = (
  id?: string,
  threshold?: number
): ScoreData | undefined => {
  if (id === undefined) return undefined;
  const dispatch = useAppDispatch();
  const [isFetching, setIsFetching] = useState(false);
  const [scoreData, setScoreData] = useState<ScoreData>({
    scoredSongsHashes: [],
    scoresRecent: [],
    scoresTop: [],
  });

  useEffect(() => {
    setIsFetching(true);
    const fetchData = async () => {
      try {
        const resp = await toast.promise<ScoreData>(
          getAllScores(id, threshold),
          {
            pending: 'Updating scores ...',
            success: 'We updated all Scores ðŸ‘Œ',
            error: 'There has been an issue your scores ðŸ¤¯',
          }
        );

        setScoreData(resp);
        setIsFetching(false);
      } catch (error: unknown) {
        console.log(
          'ðŸš€ ~ file: useBuildScoreData.ts ~ line 34 ~ fetchData ~ error',
          error
        );
        setIsFetching(false);
      }
    };

    void fetchData();
  }, [id]);

  dispatch(userIsFetchingData({ status: isFetching }));

  return scoreData;
};

export default useBuildScoreData;
