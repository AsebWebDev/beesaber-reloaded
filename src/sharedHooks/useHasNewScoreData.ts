import { skipToken } from '@reduxjs/toolkit/dist/query';
import { useEffect, useState } from 'react';

import { useGetFullPlayerQuery } from '@/api/services/apiPlayer/apiPlayer';
import { useGetUserDataQuery } from '@/api/services/apiUser/apiUser';

const useHasNewScoreData = (userId: string): boolean => {
  const [scoreSaberId, setScoreSaberId] = useState<string | undefined>(
    undefined
  );
  const [hasNewScores, setHasNewScores] = useState<boolean>(false);

  const { data } = useGetUserDataQuery(userId);

  const _scoreSaberId = data?.playerInfo?.playerId;
  const totalPlayCount = data?.totalPlayCount;

  useEffect(() => {
    if (_scoreSaberId !== undefined) setScoreSaberId(_scoreSaberId);
  }, [_scoreSaberId]);

  const { data: scoreSaberData } = useGetFullPlayerQuery(
    scoreSaberId ?? skipToken
  );

  useEffect(() => {
    const totalPlayCountSS = scoreSaberData?.scoreStats.totalPlayCount;

    if (
      totalPlayCount !== undefined &&
      totalPlayCountSS !== undefined &&
      totalPlayCount < totalPlayCountSS
    )
      setHasNewScores(true);
  }, [scoreSaberData]);

  return hasNewScores;
};

export default useHasNewScoreData;
