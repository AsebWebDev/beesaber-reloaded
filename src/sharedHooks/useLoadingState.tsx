import { skipToken } from '@reduxjs/toolkit/dist/query/react';

import { useGetUserDataQuery } from '@/api/services/apiUser/apiUser';
import { useAppSelector } from '@/store/hooks';
import { selectUserId } from '@/store/reducer/userDataReducer';

const useLoadingState = (): boolean => {
  const userId = useAppSelector(selectUserId);
  const { isFetching: isFetchingUserData } = useGetUserDataQuery(
    userId ?? skipToken
  );

  return isFetchingUserData;
};

export default useLoadingState;
