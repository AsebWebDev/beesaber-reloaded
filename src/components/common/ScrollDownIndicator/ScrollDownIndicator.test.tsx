import { render } from '@testing-library/react';

import * as useIsScrollingDown from '@/sharedHooks/useIsScrollingDown';

import ScrollDownIndicator from './ScrollDownIndicator';

describe('ScrollDownIndicator', () => {
  const spy = jest.spyOn(useIsScrollingDown, 'default');

  afterAll(() => {
    spy.mockReset();
  });

  it('should match the snapshot', () => {
    const { container } = render(<ScrollDownIndicator />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should return null when scrollig down', () => {
    spy.mockReturnValue({ isScrollingDown: true });

    const { container } = render(<ScrollDownIndicator />);

    expect(container.firstChild).toBeNull();
  });
});
