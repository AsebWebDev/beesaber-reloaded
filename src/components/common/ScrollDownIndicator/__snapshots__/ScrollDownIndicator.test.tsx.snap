// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollDownIndicator should match the snapshot 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

.c0::before {
  -webkit-animation: bounce 1s ease infinite;
  animation: bounce 1s ease infinite;
  bottom: 4rem;
  color: #fff;
  content: '╲╱';
  font-size: 2rem;
  height: 1rem;
  left: 50%;
  -webkit-letter-spacing: -1px;
  -moz-letter-spacing: -1px;
  -ms-letter-spacing: -1px;
  letter-spacing: -1px;
  line-height: 4rem;
  margin-left: -3rem;
  opacity: 0.8;
  position: absolute;
  text-align: center;
  width: 6rem;
}

<div
  class="c0"
  data-testid="indicator"
/>
`;
